src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+      GOTO (exit);
+   }
+
+   MONITOR_LOG (server_monitor, "polling ismaster");
+   awaited = false;
+   _server_monitor_heartbeat_started (server_monitor, awaited);
+   ret = _server_monitor_polling_ismaster (
+      server_monitor, &ismaster_reply, &error);
+
+exit:
+   duration_us = _now_us () - start_us;
+   MONITOR_LOG (
+      server_monitor, "server check duration (us): %" PRId64, duration_us);
+
+   if (ret) {
+      int64_t rtt_ms;
+      /* rtt remains -1 if awaited. */
+      if (awaited) {
+         rtt_ms = -1;
+      } else {
+         rtt_ms = duration_us / 1000;
+      }
+      mongoc_server_description_handle_ismaster (
+         description, &ismaster_reply, rtt_ms, NULL);
+      /* Check if it is a command error ("ok":0). */
+      if (description->error.code) {
+         /* TODO CDRIVER-3696: if this is a command error after the initial
+          * handshake, do not treat this like an auth error.
+          * mongoc_server_description_handle_ismaster always processes an "ok":0
+          * error as an auth error. */
+         MONITOR_LOG_ERROR (server_monitor,
+                            "error in server reply: %s",
+                            description->error.message);
+         command_or_network_error = true;
+         _server_monitor_heartbeat_failed (
+            server_monitor, &description->error, duration_us, awaited);
+      } else {
+         _server_monitor_heartbeat_succeeded (
+            server_monitor, &ismaster_reply, duration_us, awaited);
+      }
+   } else if (*cancelled) {
+      MONITOR_LOG (server_monitor, "server monitor cancelled");
+      if (server_monitor->stream) {
+         mongoc_stream_destroy (server_monitor->stream);
+      }
+      server_monitor->stream = NULL;
+      server_monitor->more_to_come = false;
+      _server_monitor_heartbeat_failed (
+         server_monitor, &description->error, duration_us, awaited);
+   } else {
+      MONITOR_LOG_ERROR (
+         server_monitor, "network error occurred: %s", error.message);
+      command_or_network_error = true;
+      mongoc_server_description_handle_ismaster (description, NULL, -1, &error);
+      _server_monitor_heartbeat_failed (
+         server_monitor, &description->error, duration_us, awaited);
+   }
+
+   if (command_or_network_error) {
+      if (server_monitor->stream) {
+         mongoc_stream_failed (server_monitor->stream);
+      }
+      server_monitor->stream = NULL;
+      server_monitor->more_to_come = false;
+      bson_mutex_lock (&server_monitor->topology->mutex);
+      _mongoc_topology_clear_connection_pool (server_monitor->topology,
+                                              server_monitor->description->id);
+      bson_mutex_unlock (&server_monitor->topology->mutex);
+   }
+
+   bson_destroy (&ismaster_reply);
+   return description;
+}
+
+/* Request scan of a single server.
+ *
+ * Caller does not need to have topology mutex locked.
+ * Locks server monitor mutex to deliver scan_requested.
```

Comment:
Should this also have a " Caller must not hold lock" line like the following function? Curious if that applies for all cases where functions take their own locks (in case earlier functions also omitted the " Caller must not hold lock." line).